### DISCUSSION TITLE
Running async sqlalchemy queries in debugpy

### DISCUSSION THREAD

**Discussion by gospodima** [sqlalchemy/sqlalchemy#11803] [Usage Questions] [answered]
There is an [issue](https://github.com/MagicStack/asyncpg/issues/1005) on asyncpg repo, which describes the problem of trying to run async sqlalchemy queries on a breakpoint with debugpy. So on attempt to run a query during debugging I am receiving following error:

```sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.InterfaceError: <class 'asyncpg.exceptions._base.InterfaceError'>: cannot perform operation: another operation is in progress```

As a project example I can reference my FastAPI + sqlalchemy [project](https://github.com/gospodima/async-fastapi-sqlalchemy-template). In that project there is also an option to use either sqlite or postgres for pytest set up and I do not receive such issue with sqlite.

I would like to ask whether somebody has similar issues or has ideas on how to solve this.

Thank you in advance.

**Comment by gospodima** [comment]
UPD: after replacing asyncpg with psycopg 3.2.1 it works fine. So most probably this is an issue on asyncpg side. But still wondering whether there are any solutions for asyncpg

**Reply by zzzeek** [reply, marked as answer]
we have to use different APIs with asyncpg since they dont follow the DBAPI.   so it's likely something to do with how their transaction blocks and/or prepared statements work (two non-DBAPI APIs we need to use)
